fix(animations): allow animations on elements in the shadow DOM

jeripeierSBB opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number:  #25672


## What is the new behavior?

enhanced containsElement method to not only use the native node contains method but to walk up the DOM tree by also considering shadow elements. Fixing this allows animations to be played if they are placed inside a shadow DOM. 

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

jeripeierSBB committed:
fix(animations): allow animations on elements in the shadow DOM

When determining whether to run an animation, the `TransitionAnimationPlayer`
checks to see if a DOM element is attached to the document. This is done by
checking to see if the element is "contained" by the document body node.

Previously, if the element was inside a shadow DOM, the engine would
determine that the element was not attached, even if the shadow DOM's
host was attached to the document. This commit updates the `containsElement()`
method on `AnimationDriver` implementations to also include shadow DOM
elements as being contained if their shadow host element is contained.

Further, when using CSS keyframes to trigger animations, the styling
was always added to the `head` element of the document, even for
animations on elements within a shadow DOM. This meant that those
elements never receive those styles and the animation would not run.
This commit updates the insertion of these styles so that they are added,
to the element's "root node", which is the nearest shadow DOM host, or the
`head` of the document if the element is not in a shadow DOM.

Closes #25672

LinPoLen commented:
ping @IgorMinar

IonelLupu commented:
hey. How can we help with this PR? Maybe we can have it in the next Angular release. It's blocking some other animation related issues

petebacondarwin commented:
I am aware that I have still not reviewed this PR. I will prioritise it for next week.

petebacondarwin reviewed (changes requested):
OK, so I finally got round to taking a look at this PR.
Thank you @jeripeierSBB for putting this together, and for your patience, everyone.
I think it is in pretty good shape. Just a couple of suggestions.

Also, since the commit message will appear in the CHANGELOG, can we reword it slightly...

```
fix(animations): allow animations on elements in the shadow DOM

When determining whether to run an animation, the `TransiationAnimationPlayer`
checks to see if a DOM element is attached to the document. This is done by
checking to see if the element is "contained" by the document body node.

Previously, if the element was inside a shadow DOM, the engine would
determine that the element was not attached, even if the shadow DOM's
host was attached to the document. This commit updates the `containsElement()`
method on `AnimationDriver` implementations to also include shadow DOM
elements as being contained if their shadow host element is contained.

Further, when using CSS keyframes to trigger animations, the styling
was always added to the `head` element of the document, even for
animations on elements within a shadow DOM. This meant that those
elements never receive those styles and the animation would not run.
This commit updates the insertion of these styles so that they are added,
to the element's "root node", which is the nearest shadow DOM host, or the
`head` of the document if the element is not in a shadow DOM.

Closes #25672
```

jeripeierSBB commented:
@petebacondarwin thanks a lot for your review. I will update the PR in the next days.

jeripeierSBB commented on packages/animations/browser/test/render/css_keyframes/css_keyframes_driver_spec.ts:
```diff
@@ -396,3 +422,7 @@ function isIE() {
   return (window as any).document['documentMode'] ? true : false;
 }
+
+function supportsShadowDOM(): boolean {
+  return typeof (<any>document.body).attachShadow !== 'undefined';
+}
```
Changed to use BrowserDetection from @angular/platform-browser/testing/src/browser_util

petebacondarwin replied:
Good call. Thanks

petebacondarwin reviewed (approved):
LGTM! Thanks @jeripeierSBB 

Reviewed-for: size-tracking

We will need to get further reviews

crisbeto commented on packages/animations/browser/src/render/css_keyframes/css_keyframes_driver.ts:
```diff
@@ -118,6 +118,18 @@ export class CssKeyframesDriver implements AnimationDriver {
+// TODO: Once we drop IE 11 support, method can be simplified
+// to the native browser function `getRootNode`
+function findNodeToAppendKfElm(element: any): Node {
+  while (element && element !== document.documentElement) {
```
Maybe I'm missing some context, but wouldn't this loop always resolve to `document.documentElement` if `shadowRoot` isn't supported? [According to MDN](https://developer.mozilla.org/en-US/docs/Web/API/Element/shadowRoot), IE doesn't support `shadowRoot` either. 

If that's the case, we could get away with something like `element.getRootNode?.() || document.head`.

jeripeierSBB replied:
I think you are right, my original thought (back in December ;) ) was that ShadowRoot was introduced a few versions earlier in some browsers than getRootNode. But for example Firefox had getRootNode before ShadowRoot... So I think this doesn't matter.
Shadow Dom doesn't seem to be supported in Internet Explorer without a polyfill.

crisbeto commented on packages/animations/browser/src/render/css_keyframes/css_keyframes_driver.ts:
```diff
@@ -118,6 +118,18 @@ export class CssKeyframesDriver implements AnimationDriver {
+    }
+    element = element.parentNode || element.host;
+  }
+  return document.querySelector('head')!;
```
You should be able to use `document.head` for this.

crisbeto commented on packages/animations/browser/src/render/css_keyframes/css_keyframes_driver.ts:
```diff
@@ -118,6 +118,18 @@ export class CssKeyframesDriver implements AnimationDriver {
 
+// TODO: Once we drop IE 11 support, method can be simplified
+// to the native browser function `getRootNode`
+function findNodeToAppendKfElm(element: any): Node {
```
`findNodeToAppendKfElm` -> `findNodeToAppendKeyframeElmement`. We should avoid abbreviations in symbol names.

crisbeto commented on packages/animations/browser/src/render/css_keyframes/css_keyframes_driver.ts:
```diff
@@ -118,6 +118,18 @@ export class CssKeyframesDriver implements AnimationDriver {
+    }
+    element = element.parentNode || element.host;
+  }
+  return document.querySelector('head')!;
```
Also do we need to account for non-browser environments here?

jeripeierSBB replied:
I can't decide that, but I think if using your suggested shortcut (getRootNode), we could not save a lot if checking for isBrowser and creating a switch. Non browser environments would fallback to document.head (as it was before anyways).

jeripeierSBB committed:
fixup! fix(animations): allow animations on elements in the shadow DOM

When determining whether to run an animation, the `TransitionAnimationPlayer`
checks to see if a DOM element is attached to the document. This is done by
checking to see if the element is "contained" by the document body node.

Previously, if the element was inside a shadow DOM, the engine would
determine that the element was not attached, even if the shadow DOM's
host was attached to the document. This commit updates the `containsElement()`
method on `AnimationDriver` implementations to also include shadow DOM
elements as being contained if their shadow host element is contained.

Further, when using CSS keyframes to trigger animations, the styling
was always added to the `head` element of the document, even for
animations on elements within a shadow DOM. This meant that those
elements never receive those styles and the animation would not run.
This commit updates the insertion of these styles so that they are added,
to the element's "root node", which is the nearest shadow DOM host, or the
`head` of the document if the element is not in a shadow DOM.

Closes #25672

crisbeto commented on packages/animations/browser/src/render/shared.ts:
```diff
@@ -161,9 +161,19 @@ let _query: (element: any, selector: string, multi: boolean) => any[] =
-  _contains = (elm1: any, elm2: any) => {
-    return elm1.contains(elm2) as boolean;
-  };
+  if (!isBrowser()) {
```
Isn't this check always going to be true? There's an `if (_isNode)` a couple of lines above.

jeripeierSBB replied:
No because, `_isNode || typeof Element !== 'undefined'` is true in browsers and node applications (`typeof Element !== 'undefined'`).
Do you know why the check of `_isNode || typeof Element !== 'undefined'` exists? This seems to match in almost all circumstances.

The reason I made the isBrowser check was to introduce a shortcut to not check for shadow hosts in non browser environments. But I think we could in every case also use the part of the else block and remove the switch. What do you think?

crisbeto replied:
I see. Let's keep it as is then.

crisbeto reviewed (approved):
LGTM

petebacondarwin commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -12,7 +12,7 @@
       "uncompressed": {
         "runtime-es2015": 3033,
-        "main-es2015": 451600,
+        "main-es2015": 452289,
```
@jeripeierSBB this can no be changed to `452203` since you reduced the `findNodeToAppendKeyframeElement()` function size.

petebacondarwin commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -21,7 +21,7 @@
       "uncompressed": {
         "runtime-es2015": 3153,
-        "main-es2015": 437306,
+        "main-es2015": 437924,
```
Similarly this one can be `437844`.

jeripeierSBB replied:
Sorry I did not really get the size tracking thing.. Thanks for pointing me in the right direction ;)

jeripeierSBB replied:
@petebacondarwin I tried several ways to locally get the right size to enter here, but failed. What is the correct way to get the bundle file size which should be entered here (not waiting for ci to fail)?

petebacondarwin reviewed (approved):
One last change and then we are good to merge.

jeripeierSBB committed:
fixup! fix(animations): allow animations on elements in the shadow DOM

When determining whether to run an animation, the `TransitionAnimationPlayer`
checks to see if a DOM element is attached to the document. This is done by
checking to see if the element is "contained" by the document body node.

Previously, if the element was inside a shadow DOM, the engine would
determine that the element was not attached, even if the shadow DOM's
host was attached to the document. This commit updates the `containsElement()`
method on `AnimationDriver` implementations to also include shadow DOM
elements as being contained if their shadow host element is contained.

Further, when using CSS keyframes to trigger animations, the styling
was always added to the `head` element of the document, even for
animations on elements within a shadow DOM. This meant that those
elements never receive those styles and the animation would not run.
This commit updates the insertion of these styles so that they are added,
to the element's "root node", which is the nearest shadow DOM host, or the
`head` of the document if the element is not in a shadow DOM.

Closes #25672

alxhub commented on packages/animations/browser/src/render/shared.ts:
```diff
@@ -161,9 +161,19 @@ let _query: (element: any, selector: string, multi: boolean) => any[] =
+    _contains = (elm1, elm2) => elm1.contains(elm2);
+  } else {
+    _contains = (elm1, elm2) => {
+      while (elm2 && elm2 !== document.documentElement) {
```
Would it be worth having a short circuit `if (elm1.contains(elm2)) { return true; }` here to optimize the common case?

jeripeierSBB replied:
I guess that ‚contains‘ internally also does kind of tree walking, so it would not speed up a lot but slow down by 2 in the shadow dom context usage

atscott reviewed (approved):
Reviewed-for: size-tracking

AndrewKushnir commented on packages/animations/browser/src/render/shared.ts:
```diff
@@ -161,9 +161,19 @@ let _query: (element: any, selector: string, multi: boolean) => any[] =
 const _isNode = isNode();
 if (_isNode || typeof Element !== 'undefined') {
   // this is well supported in all browsers
```
I'm a bit confused by this comment, should we remove/update it as well?

jeripeierSBB replied:
I will remove the comment as it does not provide any information.

jeripeierSBB committed:
fixup! fix(animations): allow animations on elements in the shadow DOM

When determining whether to run an animation, the `TransitionAnimationPlayer`
checks to see if a DOM element is attached to the document. This is done by
checking to see if the element is "contained" by the document body node.

Previously, if the element was inside a shadow DOM, the engine would
determine that the element was not attached, even if the shadow DOM's
host was attached to the document. This commit updates the `containsElement()`
method on `AnimationDriver` implementations to also include shadow DOM
elements as being contained if their shadow host element is contained.

Further, when using CSS keyframes to trigger animations, the styling
was always added to the `head` element of the document, even for
animations on elements within a shadow DOM. This meant that those
elements never receive those styles and the animation would not run.
This commit updates the insertion of these styles so that they are added,
to the element's "root node", which is the nearest shadow DOM host, or the
`head` of the document if the element is not in a shadow DOM.

Closes #25672

AndrewKushnir commented:
[Presubmit](http://test/OCL:368264198:BASE:368264220:1618340234704:884a7b3).

AndrewKushnir commented:
@petebacondarwin FYI, presubmits (including the global one) are successful for the changes in this PR.

petebacondarwin commented:
Thanks @AndrewKushnir - let's merge it!

AndrewKushnir commented:
FYI, this change was merged into the `master` and `12.0.x` (RC) branches, but there was a conflict while merging into `11.2.x` branch. If the change should also be included into the `11.2.x` branch, please create a new PR targeting `11.2.x` branch. Thank you.

lubos-bistak commented:
Is it possible to apply this fix also to Angular 9?

petebacondarwin commented:
Since v9 is in LTS, it will not receive this bug fix, sorry.
